### ISSUE TITLE
Manage Document lifecycle

### ISSUE THREAD

**Issue by quartzmo** [GoogleCloudPlatform/gcloud-ruby#417]
Provide initialization, create, and delete for Documents. Document update is not supported. Documents can be updated by creating a new Document with the same document id, which will replace the existing document without error or warning.

- [ ] Initialize a local, unsaved document instance, which is not yet added to its index on the server. Providing a unique document id (as shown below) is optional. It will be generated by the service if not provided.

```ruby
require "gcloud"

gcloud = Gcloud.new
search = gcloud.search

index = search.index "products"
document = index.document "product-sku-000001"
index.exists? document # API call
#=> false
document.rank #=> nil
document.id #=> "product-sku-000001"
```

- [ ] Save the document into the index. This is a separate lifecycle step from initialization so that the document's fields can be set. Saving with [documents.create](https://cloud.google.com/search/reference/rest/v1/projects/indexes/documents/create) should populate the Document's `rank` attribute from the representation returned by the service. Since update is not supported, it should be allowed to call `create` on a document that already exists.

```ruby
document["title"] << "Hello Gcloud!" # Give the document its data.

index.save document  # API call
index.exists? document # API call
#=> true
document.rank      # set by the server
#=> 1443648166
```

- [ ] Delete a document from its index with [documents.delete](https://cloud.google.com/search/reference/rest/v1/projects/indexes/documents/delete):

```ruby
require "gcloud"

gcloud = Gcloud.new
search = gcloud.search

document = index.document "product-sku-000001"
index.exists? document # API call
#=> true
index.delete document  # API call
index.exists? document # API call
#=> false
```

To update a document in place after manipulating its fields or rank, just recreate it: E.g.:

```ruby
require "gcloud"

gcloud = Gcloud.new
search = gcloud.search

document = index.document "product-sku-000001"
index.exists? document # API call
#=> true
document.rank = 12345 # Set a new rank 
document["title"] << "Hello World" # Update the document its data
index.save document  # API call
```

**Comment by blowmage** [comment]
Looking at the API a bit more, specifically the [Document data structure](https://cloud.google.com/search/reference/rest/v1/projects/indexes/documents#resource_representation.google.cloudsearch.v1.Document), I think we should make the document object a value object. The reason for this is that the document data does not know what project or index it belongs to. It would be possible to store the document data in multiple indexes, across multiple projects, unchanged. So I think Search Document should behave like a Datastore Entity, and the Search Index should behave like Datastore Dataset. Repository pattern and all that. :)

```ruby
search = gcloud.search
index = search.index "products"
document = Gcloud::Search::Document.new
document.doc_id = "happy-ruby-sku-000001"
document.fields = { ... }
index.save document
```

**Comment by quartzmo** [comment]
This makes sense to me. I'll update the description examples.

**Comment by quartzmo** [comment]
Just a note, we still need `Index#document` (or perhaps a new method, `Index#create_document`?) to create new, unsaved documents, and tests for this.

**Comment by blowmage** [comment]
Yeah, on Datastore we have Dataset#entity that is equivalent to Entity.new. Since Search is also following the Repository pattern, we may consider following those renaming Index#document to Index#find, and then make Index#document a shortcut for Document.new.

**Comment by quartzmo** [comment]
Sounds fine to me.

**Comment by quartzmo** [comment]
Closed by #427, #440.
